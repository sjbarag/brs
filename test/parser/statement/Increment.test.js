const brs = require("brs");
const { Lexeme } = brs.lexer;
const { Int32, BrsBoolean } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("parser postfix unary expressions", () => {
    let parser;

    beforeEach(() => {
        parser = new brs.parser.Parser();
    });

    it("parses postfix '++' for variables", () => {
        let { statements, errors } = parser.parse([
            identifier("foo"),
            token(Lexeme.PlusPlus, "++"),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses postfix '--' for dotted get expressions", () => {
        let { statements, errors } = parser.parse([
            identifier("obj"),
            token(Lexeme.Dot, "."),
            identifier("property"),
            token(Lexeme.MinusMinus, "--"),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("parses postfix '++' for indexed get expressions", () => {
        let { statements, errors } = parser.parse([
            identifier("obj"),
            token(Lexeme.LeftSquare, "["),
            identifier("property"),
            token(Lexeme.RightSquare, "]"),
            token(Lexeme.PlusPlus, "++"),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("disallows consecutive postfix operators", () => {
        let { errors } = parser.parse([
            identifier("foo"),
            token(Lexeme.PlusPlus, "++"),
            token(Lexeme.PlusPlus, "++"),
            EOF
        ]);

        expect(errors).toHaveLength(1);
        expect(errors[0]).toMatchObject({
            message: expect.stringMatching("Consecutive increment/decrement operators are not allowed")
        });
    });

    it("disallows postfix '--' for function call results", () => {
        let { errors } = parser.parse([
            identifier("func"),
            token(Lexeme.LeftParen, "("),
            token(Lexeme.RightParen, ")"),
            token(Lexeme.MinusMinus, "--"),
            EOF
        ]);

        expect(errors).toHaveLength(1);
        expect(errors[0]).toMatchObject({
            message: expect.stringMatching("Increment/decrement operators are not allowed on the result of a function call")
        });
    });

    it("allows '++' at the end of a function", () => {
        let { statements, errors } = parser.parse([
            token(Lexeme.Sub, "sub"),
            identifier("foo"),
            token(Lexeme.LeftParen, "("),
            token(Lexeme.RightParen, ")"),
            token(Lexeme.Newline, "\n"),
            identifier("someValue"),
            token(Lexeme.PlusPlus, "++"),
            token(Lexeme.Newline, "\n"),
            token(Lexeme.EndSub, "end sub"),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    test("location tracking", () => {
        /**
         *    0   0   0   1
         *    0   4   8   2
         *  +--------------
         * 1| someNumber++
         */
        let { statements, errors } = parser.parse([
            {
                kind: Lexeme.Identifier,
                text: "someNumber",
                isReserved: false,
                location: {
                    start: { line: 1, column: 0 },
                    end: { line: 1, column: 10 }
                }
            },
            {
                kind: Lexeme.PlusPlus,
                text: "++",
                isReserved: false,
                location: {
                    start: { line: 1, column: 10 },
                    end: { line: 1, column: 12 },
                }
            },
            {
                kind: Lexeme.Eof,
                text: "\0",
                isReserved: false,
                location: {
                    start: { line: 1, column: 12 },
                    end: { line: 1, column: 13 }
                }
            }
        ]);

        expect(errors).toEqual([]);
        expect(statements).toHaveLength(1);
        expect(statements[0].location).toMatchObject({
            start: { line: 1, column: 0 },
            end: { line: 1, column: 12 }
        });
    });
});