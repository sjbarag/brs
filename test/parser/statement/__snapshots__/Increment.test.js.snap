// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser postfix unary expressions allows '++' at the end of a function 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [
          Increment {
            "token": Object {
              "isReserved": false,
              "kind": "PlusPlus",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": -9,
                  "line": -9,
                },
                "start": Object {
                  "column": -9,
                  "line": -9,
                },
              },
              "text": "++",
            },
            "type": "Increment",
            "value": Variable {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "someValue",
              },
              "type": "Variable",
            },
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "foo",
    },
    "type": "Function",
  },
]
`;

exports[`parser postfix unary expressions parses postfix '++' for indexed get expressions 1`] = `
Array [
  Increment {
    "token": Object {
      "isReserved": false,
      "kind": "PlusPlus",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "++",
    },
    "type": "Increment",
    "value": IndexedGet {
      "closingSquare": Object {
        "isReserved": false,
        "kind": "RightSquare",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "]",
      },
      "index": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "property",
        },
        "type": "Variable",
      },
      "obj": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "obj",
        },
        "type": "Variable",
      },
      "type": "IndexedGet",
    },
  },
]
`;

exports[`parser postfix unary expressions parses postfix '++' for variables 1`] = `
Array [
  Increment {
    "token": Object {
      "isReserved": false,
      "kind": "PlusPlus",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "++",
    },
    "type": "Increment",
    "value": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "foo",
      },
      "type": "Variable",
    },
  },
]
`;

exports[`parser postfix unary expressions parses postfix '--' for dotted get expressions 1`] = `
Array [
  Increment {
    "token": Object {
      "isReserved": false,
      "kind": "MinusMinus",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "--",
    },
    "type": "Increment",
    "value": DottedGet {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "property",
      },
      "obj": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "obj",
        },
        "type": "Variable",
      },
      "type": "DottedGet",
    },
  },
]
`;
